flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/nn/QuadTree.scala

```diff
@@ -0,0 +1,305 @@
+      *
+      * @return
+      */
+    def getCenterLength(): (ListBuffer[Double], ListBuffer[Double]) = {
+      (c, L)
+    }
+
+    def contains(obj: Vector): Boolean = {
+      overlap(obj, 0.0)
+    }
+
+    /** Tests if obj is within a radius of the node
+      *
+      * @param obj
+      * @param radius
+      * @return
+      */
+    def overlap(obj: Vector, radius: Double): Boolean = {
+      var count = 0
+      for (i <- 0 to obj.size - 1) {
+        if (obj(i) - radius < c(i) + L(i) / 2 && obj(i) + radius > c(i) - L(i) / 2) {
+          count += 1
+        }
+      }
+
+      if (count == obj.size) {
+        return true
+      } else {
+        return false
+      }
+    }
+
+    /** Tests if obj is near a node:  minDist is defined so that every point in the box
+      * has distance to obj greater than minDist
+      * (minDist adopted from "Nearest Neighbors Queries" by N. Roussopoulos et al.)
+      *
+      * @param obj
+      * @param radius
+      * @return
+      */
+    def isNear(obj: Vector, radius: Double): Boolean = {
+      if (minDist(obj) < radius) {
+        true
+      } else {
+        false
+      }
+    }
+
+    def minDist(obj: Vector): Double = {
+      var minDist = 0.0
+      for (i <- 0 to obj.size - 1) {
+        if (obj(i) < c(i) - L(i) / 2) {
+          minDist += math.pow(obj(i) - c(i) + L(i) / 2, 2)
+        } else if (obj(i) > c(i) + L(i) / 2) {
+          minDist += math.pow(obj(i) - c(i) - L(i) / 2, 2)
+        }
+      }
+      return minDist
+    }
+
+    def whichChild(obj:Vector):Int = {
+
+      var count = 0
+      for (i <- 0 to obj.size - 1){
+        if (obj(i) > c(i)) {
+          count += Math.pow(2,i).toInt
+        }
+      }
+      count
+    }
+
+    def makeChildren() {
+      var cBuff = new ListBuffer[ListBuffer[Double]]
+      cBuff += c
+      var Childrennodes = new ListBuffer[Node]
+      val cPart = partitionBox(cBuff,L,L.length)
+      for (i <- cPart.indices){
+        Childrennodes = Childrennodes :+ new Node(cPart(i), L.map(x => x/2.0), null)
+
+      }
+      children = Childrennodes.clone()
+    }
+
+    /**
+      * Recursive function that partitions a n-dim box by taking the (n-1) dimensional
+      * plane through the center of the box keeping the n-th coordinate fixed,
+      *  then shifting it in the n-th direction up and down
+      * and recursively applying partitionBox to the two shifted (n-1) dimensional planes.
+     *
+     * @param cPart
+     * @param L
+     * @param dim
+     * @return
+     */
+    def partitionBox(cPart:ListBuffer[ListBuffer[Double]],L:ListBuffer[Double], dim:Int):
+    ListBuffer[ListBuffer[Double]]=
+    {
+      if (L.length == 1){
+
+        var cPartDown = cPart.clone()
+        //// shift center up and down
+        val cPartUp = cPart.map{v => v.patch(dim-1, Seq(v(dim - 1) + L(dim-1)/4), 1)}
+        cPartDown = cPartDown.map{v => v.patch(dim-1, Seq(v(dim - 1) - L(dim-1)/4), 1)}
+
+        return cPartDown ++ cPartUp
```

Comment:
returns are not very scalaesque. Why not putting the following code in an `else` branch?